### PR TITLE
on kill segment, dont leave version, interval and dataSource dir behind on HDFS

### PR DESCRIPTION
**Pull request by himanshug**

### CODE REVIEW

File: extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java [hunk 1]
```diff
@@ -55,6 +72,16 @@ public void kill(DataSegment segment) throws SegmentLoadingException
     }
   }
 
+  private boolean safeNonRecursiveDelete(FileSystem fs, Path path)
+  {
+    try {
+      return fs.delete(path, false);
+    }
+    catch (Exception ex) {
+      return false;
```
Comment (fjy) [inline extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java]: should we log something?

File: extensions/hdfs-storage/src/test/java/io/druid/storage/hdfs/HdfsDataSegmentKillerTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+{
+  @Test
+  public void testKill() throws Exception
+  {
+    Configuration config = new Configuration();
+    HdfsDataSegmentKiller killer = new HdfsDataSegmentKiller(config);
+
+    FileSystem fs = FileSystem.get(config);
+
+    // Create following segments and then delete them in this order and assert directory deletions
```
Comment (fjy) [inline extensions/hdfs-storage/src/test/java/io/druid/storage/hdfs/HdfsDataSegmentKillerTest.java]: can we remove this code if it is not being used?

File: extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java (hunk 1 above)
Comment (himanshug) [inline extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java]: not really, this is used to delete version, interval etc directories and it is legit for those deletions to fail as they might not be empty. FileSystem documentation about exactly what cases throw exception is not that clear so it is probably ok for failure be considered valid.

File: extensions/hdfs-storage/src/test/java/io/druid/storage/hdfs/HdfsDataSegmentKillerTest.java (hunk 2 above)
Comment (himanshug) [inline extensions/hdfs-storage/src/test/java/io/druid/storage/hdfs/HdfsDataSegmentKillerTest.java]: do you mean, remove this comment? I thought it was useful to explain what test was doing.

Comment (fjy) [inline extensions/hdfs-storage/src/test/java/io/druid/storage/hdfs/HdfsDataSegmentKillerTest.java]: oh wait, i thought this was commented out code, not an actual code comment, my bad

**Comment by fjy** [comment]
:+1:

**Comment by himanshug** [comment]
@fjy thanks for the review.
@gianm I thought more about it and realized that the "assumption" about the hierarchy has to be made in order to be able to do proper cleanup. For example, on deleting partition directory, you can delete interval directory (if empty) by only assuming that interval directory is parent of the partition directory. However, I did add a comment in druid-api https://github.com/druid-io/druid-api/pull/58 and hopefully that should be OK.

**Comment by gianm** [comment]
@himanshug ok, I think that given druid-io/druid-api#58 and the fact that this delete is non-recursive, this PR is safe enough to merge.

**Comment by gianm** [comment]
:+1:

**Comment by himanshug** [comment]
@fjy @gianm since there are 2 thumbs up, can one of you please merge this?

**Pull request merged**